lua/entities/gmod_wire_expression2/core/extpp.lua

```diff
@@ -228,279 +149,155 @@ function E2Lib.ExtPP.Pass2(contents)
-			local lookup = {}
-			for _, tag in ipairs(attributes) do
-				local k = tag:lower():Trim()
-
-				if k:find("=", 1, true) then
-					-- [xyz = 567, event = "Tick"]
-					-- e2function number foo()
-					local key, value = unpack( k:Split("="), 1, 2 )
-					key, value = key:lower():Trim(), tonumber(value:match("%d+")) or value:Trim()
-
-					lookup[key] = value
-				else
-					if not valid_attributes[k] then
-						ErrorNoHalt("Invalid attribute fed to ExtPP: " .. k)
-					end
-					lookup[ tag:lower():Trim() ] = "true"
-				end
-			end
-
-			local buf = "{"
-			for attr, val in pairs(lookup) do
-				buf = buf .. "['" .. attr .. "'] = " .. val .. ","
-			end
-
-			attributes = buf .. "}"
+		local line = select(2, contents:sub(1, h_begin):gsub("\n", ""))
+		local trace = "(at line " .. line .. ")"
+		local trace_ext = trace .. (E2Lib.currentextension and (" @" .. filename) or "")
+
+		if contents:sub(h_begin - 1, h_begin - 1):match("%w") then
+			ErrorNoHalt("Warning: Malformed e2function must not have characters before 'e2function' " .. trace_ext .. "\n")
+		elseif RemovedOperators[name] then -- Old operator that no longer is needed.
+			ErrorNoHalt("Warning: Operator " .. name .. " is now redundant. Ignoring registration. " .. trace_ext .. "\n")
+		elseif not name:find("^" .. p_funcname .. "$") and not Operators[name] then
+			error("PP syntax error: Invalid function name format " .. trace)
+		elseif thistype ~= "" and colon == "" then
+			error("PP syntax error: Function names may not start with a number. " .. trace)
+		elseif thistype == "" and colon ~= "" then
+			error("PP syntax error: No type for 'this' given." .. trace)
+		elseif thistype ~= "" and not getTypeId(thistype) then
+			error("PP syntax error: Invalid type for 'this': '" .. thistype .. "' " .. trace)
+		elseif thistype:match("^%d") then
+			error("PP syntax error: Type names may not start with a number." .. trace)
+		elseif equals ~= "" and aliaspos then
+			error("PP syntax error: Malformed alias definition. " .. trace)
+		elseif ret ~= "void" and not getTypeId(ret) then
+			error("PP syntax error: Invalid return type: '" .. ret .. "' " .. trace)
 		else
-			attributes = "{}"
-		end
+			changed = true -- Mark as changed
 
-		changed = true
-
-		local function handle_function()
-			if contents:sub(h_begin - 1, h_begin - 1):match("%w") then return end
-			local aliasflag = nil
-			if equals == "" then
-				if aliaspos then
-					if contents:sub(aliaspos, h_begin - 1):find("%S") then error("PP syntax error: Malformed alias definition.", 0) end
-					-- right hand side of an alias assignment
-					aliasflag = 2
-					aliaspos = nil
-				end
-			else
-				if aliaspos then error("PP syntax error: Malformed alias definition.", 0) end
-				-- left hand side of an alias assignment
+			local aliasflag
+			if equals == "" and aliaspos then -- Alias right hand side
+				assert(not contents:sub(aliaspos, h_begin - 1):find("%S"), "PP syntax error: Malformed alias definition. " .. trace)
```

Comment:
assert is for asserting invariants, not reporting errors.
Also, this is hard to read with that negation.

Reply:
The assert is fine. The negation could be removed by changing the pattern match. Maybe it could just do `s:Trim() == ""` or `s:match("^%s*$")`. Readability isn't much of a priority for this file since nobody should be changing it very often (and since I want to remove it for a better api). If we did care that much about it the gigantic pattern match would be removed for an actual parser :\

Reply:
this is just a side point anyway. the reason I commented is that your change made it less readable, seemingly for no reason other than using the assert function more (which is not a good reason)

This PR is hard to review if you only make the necessary changes.
If you add unnecessary changes to it, that'll make it even harder, so at least get rid of the unnecessary changes that make the code less readable.

Reply:
Nah I'll probably get to it I didn't really like how it looked either, probably just left it like that cause I didn't want to refactor the logic (just using the existing error assertion)